### PR TITLE
Migrate to CircleCI2.0 and Add AppVeyor for master-only branch

### PR DESCRIPTION
**Pull request by raphamorim**
Ref: #11595 

Result of AppVeyor on my fork: https://ci.appveyor.com/project/raphamorim/react/build/1.0.30
Result of CircleCI 2.0 on my fork: https://circleci.com/gh/raphamorim/react/21

cc// @gaearon, @felicianotech

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    steps:
+      - checkout
+
+      - run: echo $CIRCLE_COMPARE_URL | cut -d/ -f7
+
+      - run:
+          name: Install Yarn
+          command: |
+            if [[ ! -e ~/.yarn/bin/yarn || $(yarn --version) != "${YARN_VERSION}" ]]; then
+              curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version $YARN_VERSION
```
Comment (jquense) [inline .circleci/config.yml]: any reason to pin the version? docker/node comes with at least yarn 1.2.1 already

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+version: 2
+jobs:
+  build:
+
+    docker:
+      - image: circleci/node:8
```
Comment (jquense) [inline .circleci/config.yml]: worth using `8-alpine`?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      - run:
+          name: Install Packages
+          command: yarn install
+
+      - run:
+          name: Test Packages
+          command: ./scripts/circleci/test_entry_point.sh
+
+      - save_cache:
+          key: node-modules-{{ checksum "yarn.lock" }}
```
Comment (jquense) [inline .circleci/config.yml]: probably worth prefixing/appending `-v1` to the key or something so it can be neatly manually invalidated if required

**Comment by jquense** [comment]
lsorry for all the notes :) i realize that most of those things are a straight port from the v1 config, but worth thinking about while the pieces are being moved

File: .circleci/config.yml (hunk 1 above)
Comment (raphamorim) [reply .circleci/config.yml]: I guess `circleci/node:8` uses `yarn@1.3.1`

File: .circleci/config.yml (hunk 2 above)
Comment (raphamorim) [reply .circleci/config.yml]: I'm using `circleci` docker images,
https://hub.docker.com/r/circleci/node/tags/

Unfortunately there's none image with Alpine. But this is nice question: Move foward with `8-alpine` (https://hub.docker.com/_/node/)?

**Comment by raphamorim** [comment]
No problem at all @jquense.

> but worth thinking about while the pieces are being moved

I totally agree.

**Comment by gaearon** [comment]
I created an AppVeyor project. What's the next step?

**Comment by raphamorim** [comment]
@gaearon it's running only for `master` branch. 
Check the history of latest builds. This error is because `appveyor.yml` should set `build: off`.

https://ci.appveyor.com/project/gaearon/react/history

**Comment by raphamorim** [comment]
The next step is configure CircleCI to migrate for 2.0. This should be expecting for 2.0 common setting files.

**Comment by raphamorim** [comment]
The strangest thing is triggering on my personal CI: https://circleci.com/gh/raphamorim/react/26
Not the facebook/react.

**Comment by raphamorim** [comment]
Note: Remove "Upload build" on CircleCi 2.0 config in favour of #11666

File: circle.yml [hunk 4]
```diff
@@ -1,34 +0,0 @@
...
-        curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version $YARN_VERSION
-      fi
-  override:
-    - yarn install
-    - scripts/circleci/set_up_github_keys.sh
-  post:
-    # - npm ls --depth=0
-  cache_directories:
-    - ~/.yarn
-    - ~/.yarn-cache
```
Comment (gaearon) [inline circle.yml]: Are these irrelevant now?

**Comment by raphamorim** [comment]
Yup @gaearon.

I made some changes based on official docs about [Caching for 2.0](https://circleci.com/docs/2.0/caching/) and reading [yarn package CircleCI configuration code](https://github.com/yarnpkg/yarn/blob/master/.circleci/config.yml).

Note the warning on 1.0 build (e.g: https://circleci.com/gh/facebook/react/7853#tests/containers/3):
`circle.yml specified cache directory: /home/ubuntu/.yarn-cache but it does not exist`

BTW: I'll sent a different PR for add AppVeyor badge on `README.md`

File: circle.yml (hunk 4 above)
Comment (raphamorim) [reply circle.yml]: 👍  :octocat:

**Comment by gaearon** [comment]
Looks good to me. Thanks.

**Comment by felicianotech** [comment]
I so missed the notification for this, sorry. FWIW, the CircleCI config looks good. I did notice it's missing the opportunity to store test results and thus the ability to graph testing data overtime. We can discuss in a new Issue if anyone is interested, just let me know.

**Pull request merged**